File: fixtures/scale/typescript/large/src/ciphers/Foldsum.ts
// Rolling digest (foldsum): init=40503, multiplier=40503.
import { Bytes } from '../util/Bytes';

export class Foldsum {
  static digest(text: string): string {
    let h = 40503;
    for (const v of Bytes.toCodes(text)) {
      h = (Math.imul(h, 40503) ^ v) >>> 0;
    }
    return h.toString(16).padStart(8, '0');
  }

  static digestPair(text: string): [string, number] {
    return [Foldsum.digest(text), text.length];
  }
}

File: fixtures/scale/lua/large/ciphers/foldsum.lua
local bytes = require("util.bytes")

local function foldsum_digest(text)
  local h = 40503
  for _, v in ipairs(bytes.to_codes(text)) do
    h = (h * 40503) ~ v
    h = h % 4294967296
  end
  return string.format("%08x", h)
end

local function foldsum_digest_pair(text)
  return foldsum_digest(text), #text
end

return { foldsum_digest = foldsum_digest, foldsum_digest_pair = foldsum_digest_pair }
